include/seqan3/contrib/parallel/reader_writer_latch.hpp

```diff
@@ -0,0 +1,323 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::contrib::reader_writer_latch and utility objects.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <cassert>
+#include <mutex>
+
+#include <seqan3/contrib/parallel/latch.hpp>
+#include <seqan3/contrib/parallel/spin_delay.hpp>
+#include <seqan3/core/detail/strong_type.hpp>
+#include <seqan3/std/new>
+
+namespace seqan3::contrib
+{
+//!\cond
+//!\brief A strong type to set the writer count of a seqan3::contrib::reader_writer_latch.
+struct writer_count : public detail::strong_type<size_t, writer_count>
+{
+    //!\brief An alias for the base type.
+    using base_t = detail::strong_type<size_t, writer_count>;
+
+    //!\brief Import the base constructor.
+    using base_t::base_t;
+};
+
+//!\brief A strong type to set the reader count of a seqan3::contrib::reader_writer_latch.
+struct reader_count : public detail::strong_type<size_t, reader_count>
+{
+    //!\brief An alias for the base type.
+    using base_t = detail::strong_type<size_t, reader_count>;
+
+    //!\brief Import the base constructor.
+    using base_t::base_t;
+};
+
+/*!\brief A single-use synchronisation point for closable concurrent data structures.
```

Comment:
What are `closable concurrent data structures`?

Reply:
Yes, we don't have a concept for this yet. But sooner or later we are going to bring this from contrib into seqan3 I think. The proposal for concurrent queues introduces the concept of closing a queue. If a queue is closed it won't allow pushes anymore. It also allows us to check when a queue is done reading from. Before in SeqAn this was managed internally by the queue itself by knowing the number of participating readers/writers. These latches/guards are used to achieve the same thing, but as external data structures.